### DISCUSSION TITLE
Is there a "retry from error" in v2?

### DISCUSSION THREAD

**Discussion by mohebifar** [pmndrs/jotai#1826] [Q&A]
Hello,

I recently upgraded to v2 of Jotai and noticed some significant internal changes, including the removal of the `retryFromError` functionality. In the past, I relied on this feature in my `ErrorBoundary` component to re-trigger `atomWithObservable`s, but I'm having trouble finding a viable alternative in v2.

```ts
  const [retries, setRetries] = React.useState(0);
  const ScopeContext = SECRET_INTERNAL_getScopeContext();
  const { r: retryFromError } = React.useContext(ScopeContext);

  const retry = React.useCallback(() => {
    retryFromError?.(() => {
      setRetries((prev) => prev + 1);
    });
  }, [retryFromError]);


  return <ErrorBoundaryUI onRetry={retry} />
```

Would someone be able to provide some guidance or suggestions for how I can achieve similar functionality in v2? I would greatly appreciate any insights or help that can be provided.

**Comment by mohebifar** [comment]
Ok, I just came up with a workaround for now, but I'm curious if there's another way of doing this.

I saved a WeakMap of Error -> refreshAtom and then, when an error is caught, I grab the refreshAtom for that specific error and increment its value.


```ts
// getRefreshAtoms.ts
import type { PrimitiveAtom } from "jotai";

type RefreshAtom = PrimitiveAtom<number>;

const refreshAtomsMap = new WeakMap<Error, RefreshAtom>();

export const setRefreshAtom = (error: Error, refreshAtom: RefreshAtom) => {
  refreshAtomsMap.set(error, refreshAtom);
};

export const getRefreshAtom = (error: Error) => refreshAtomsMap.get(error);
```

```ts
// atom.ts
const myAtom = () => {
  const refreshAtom = atom(0);

  const subject = createObservable();

  subject.subscribe({
    error: (error) => {
      if (error instanceof Error) {
        setRefreshAtom(error, queryRefreshAtom);
      }
    },
  });

  return atomWithObservable(() => subject);
}
```

```ts
// ErrorBoundary.ts
const setRefresh = useSetAtom((error && getRefreshAtom(error)) ?? atom(0));

const retry = () => {
  setRefresh((c) => c + 1);
  reset();
};

return <ErrorBoundaryUI onRetry={retry} />
```

**Reply by wootra** [reply]
I had a similar situation. But I chose to retry only when I want retry (i.e. when some window is open that is using the value. in this case, initial call is pretty much pre-fetch). back to your issue, IMO, the approach to retry on error without limit count sounds a little risky, though. Just for reference, you might want to see my case and examples in here: https://www.npmjs.com/package/jotai-reloadable

**Comment by dai-shi** [comment]
curious if `retryFromError` in v1 was just nothing in your use case? It should have worked without it.

**Reply by mohebifar** [reply]
It actually allowed me to refresh async atoms that had failed with error/got reject.

**Reply by dai-shi** [reply]
Hmm, interesting.

```js
    retryFromError?.(() => {
      setRetries((prev) => prev + 1);
    });
```

☝️ should be equal to 👇 :

```js
    setRetries((prev) => prev + 1);
```

 as far as I recall, unless you use `unstable_versionedWrite` in v1, which is removed in v2.
